### PR TITLE
Add ServiceParameter support to the system resource

### PR DESCRIPTION
**Pull request by albailey**
- Update to the latest gophercloud.
The latest gophercloud added support for serviceparameters api.

 - Adds a new serviceParameters section to System spec.
This allows service parameters to be configured during
deployment-manager zero-touch-provisioning.

Example of updating an existing service parameter (the default was 300)

apiVersion: starlingx.windriver.com/v1
kind: System
spec:
  serviceParameters:
  - service: horizon
    section: auth
    paramname: lockout_seconds
    paramvalue: "333"

Example of adding an service parameter (note: the resource field is required)

apiVersion: starlingx.windriver.com/v1
kind: System
spec:
  serviceParameters:
  - service: platform
    section: docker
    paramname: docker_cleanup_cron_hour
    paramvalue: "20"
    resource: "platform::docker::params::docker_cleanup_cron_hour"

Signed-off-by: albailey <Al.Bailey@windriver.com>

### CODE REVIEW

File: pkg/controller/system/system_controller.go [hunk 1]
```diff
@@ -347,6 +348,102 @@ func (r *ReconcileSystem) ReconcilePTP(client *gophercloud.ServiceClient, instan
...
+					updated = true
+					r.NormalEvent(instance, common.ResourceUpdated, "ServiceParameter %q %q %q has been modified", result.Service, result.Section, result.ParamName)
+				}
+				break
+			}
+		}
+		// If a service parameter is found that does not exist in Info, need to create it
+		// Note: the sysinv API does not actually support creating a new service param.  It will be rejected.
+		if !found {
+			// TODO: populate this params object
```
Comment (donpenney) [inline pkg/controller/system/system_controller.go]: Is there something more to be done here?

File: config/crds/starlingx_v1_system.yaml [hunk 2]
```diff
@@ -149,6 +149,46 @@ spec:
...
+                      parameter
+                    type: string
+                  section:
+                    description: Section identifies the section for this service parameter
+                    maxLength: 255
+                    pattern: ^[a-zA-Z0-9\-_]+$
+                    type: string
+                  service:
+                    description: Service identifies the service for this service parameter
+                    maxLength: 255
```
Comment (mpeters-wrs) [inline config/crds/starlingx_v1_system.yaml]: Maximum length is 16.

File: config/crds/starlingx_v1_system.yaml [hunk 3]
```diff
@@ -149,6 +149,46 @@ spec:
...
+                    description: Personality identifies the personality for this service
+                      parameter
+                    type: string
+                  resource:
+                    description: Resource identifies the resource for this service
+                      parameter
+                    type: string
+                  section:
+                    description: Section identifies the section for this service parameter
+                    maxLength: 255
```
Comment (mpeters-wrs) [inline config/crds/starlingx_v1_system.yaml]: Maximum length is 128

File: pkg/apis/starlingx/v1/system_types.go [hunk 4]
```diff
@@ -91,6 +91,41 @@ func (in *LicenseInfo) DeepEqual(other *LicenseInfo) bool {
 	return other != nil
 }
 
+// ServiceParameterInfo defines the attributes required to define an instance of a
+// service parameter to be installed via the system API.
+type ServiceParameterInfo struct {
+	// Service identifies the service for this service parameter
+	// +kubebuilder:validation:Pattern=^[a-zA-Z0-9\-_]+$
+	// +kubebuilder:validation:MaxLength=255
```
Comment (mpeters-wrs) [inline pkg/apis/starlingx/v1/system_types.go]: MaxLength=16

File: pkg/apis/starlingx/v1/system_types.go [hunk 5]
```diff
@@ -91,6 +91,41 @@ func (in *LicenseInfo) DeepEqual(other *LicenseInfo) bool {
...
+	Service string `json:"service"`
+
+	// Section identifies the section for this service parameter
+	// +kubebuilder:validation:Pattern=^[a-zA-Z0-9\-_]+$
+	// +kubebuilder:validation:MaxLength=255
```
Comment (mpeters-wrs) [inline pkg/apis/starlingx/v1/system_types.go]: MaxLength=128

File: pkg/apis/starlingx/v1/system_types.go [hunk 6]
```diff
@@ -91,6 +91,41 @@ func (in *LicenseInfo) DeepEqual(other *LicenseInfo) bool {
...
+	Section string `json:"section"`
+
+	// ParamName identifies the name for this service parameter
+	// +kubebuilder:validation:Pattern=^[a-zA-Z0-9\-_]+$
+	// +kubebuilder:validation:MaxLength=255
+	ParamName string `json:"paramname"`
+
+	// ParamValue identifies the value for this service parameter
+	ParamValue string `json:"paramvalue"`
```
Comment (mpeters-wrs) [inline pkg/apis/starlingx/v1/system_types.go]: MaxLength=4096

File: config/crds/starlingx_v1_system.yaml [hunk 7]
```diff
@@ -149,6 +149,46 @@ spec:
...
+            serviceParameters:
+              description: ServiceParameters is a list of service parameters
+              items:
+                properties:
+                  paramname:
+                    description: ParamName identifies the name for this service parameter
+                    maxLength: 255
+                    pattern: ^[a-zA-Z0-9\-_]+$
+                    type: string
+                  paramvalue:
```
Comment (mpeters-wrs) [inline config/crds/starlingx_v1_system.yaml]: Maximum length is 4096

File: pkg/controller/system/system_controller.go [hunk 8]
```diff
@@ -347,6 +348,102 @@ func (r *ReconcileSystem) ReconcilePTP(client *gophercloud.ServiceClient, instan
...
+					}
+					// success
+					updated = true
+					r.NormalEvent(instance, common.ResourceUpdated, "ServiceParameter %q %q %q has been modified", result.Service, result.Section, result.ParamName)
+				}
+				break
+			}
+		}
+		// If a service parameter is found that does not exist in Info, need to create it
+		// Note: the sysinv API does not actually support creating a new service param.  It will be rejected.
```
Comment (mpeters-wrs) [inline pkg/controller/system/system_controller.go]: I think this comment note is invalid and should be removed.

File: pkg/controller/system/system_controller.go (hunk 1 above)
Comment (albailey) [reply pkg/controller/system/system_controller.go]: No.   I did eventually populate the object, but forgot to remove my todo.  I will remove the todo.

File: config/crds/starlingx_v1_system.yaml (hunk 2 above)
Comment (albailey) [reply config/crds/starlingx_v1_system.yaml]: OK, I'll change that

File: pkg/controller/system/system_controller.go (hunk 8 above)
Comment (albailey) [reply pkg/controller/system/system_controller.go]: will do

File: config/crds/starlingx_v1_system.yaml (hunk 7 above)
Comment (albailey) [reply config/crds/starlingx_v1_system.yaml]: updated

File: pkg/apis/starlingx/v1/system_types.go (hunk 6 above)
Comment (albailey) [reply pkg/apis/starlingx/v1/system_types.go]: updated

**Comment by albailey** [comment]
I have updated the max length for param_value to 4096 
I have updated section to 128, and service to 16
I also added maxlength for personality and resource based on the model.py and the migration scripts in sysinv to be 255

**Pull request merged**